exorcist/taskdb.py

```diff
@@ -198,39 +234,245 @@ def _create_empty_db(metadata, engine):
         Parameters
         ----------
         taskid: str
             task ID of the task to update
         status: TaskStatus
             the status to change to
-        old_status: TaskStatus
-            the previous status
+        is_checkout: bool
+            True if this is a checkout operation (and therefore should
+            update the number of tries)
+        max_tries: Optional[int]
+            value to set for the maximum number of trials for this task
+            (this is total tries, so retries + 1)
+        old_status: Optional[TaskStatus]
+            the previous status. If specified, the task will only match if
+            this is its current status. Default (None) allows update from
+            any previous status
+
+        Returns
+        -------
+        SQLStatement :
+            The statement to be executed to update this task row. This
+            statement will return the rows that have been changed by the
+            update.
+
+        See also
+        --------
+        _validate_update_result : validate result from statement's execution
+        """
+        stmt = (
+            sqla.update(self.tasks_table)
+            .where(self.tasks_table.c.taskid == taskid)
+        )
+
+        if old_status is not None:
+            stmt = stmt.where(self.tasks_table.c.status == old_status.value)
+
+        # create a dict of values to update
+        values = {
+            'status': status.value,
+            'last_modified': datetime.now(),
+        }
+
+        if is_checkout:
+            values['tries'] = self.tasks_table.c.tries + 1
+
+        if max_tries is not None:
+            values['max_tries'] = max_tries
+
+        stmt = stmt.values(**values)
+        return stmt
+
+    @staticmethod
+    def _validate_update_result(result):
+        """The result of an update statement should only have 1 changed row.
+
+        This gives standard validation to raise errors if things went wrong.
 
         Raises
         ------
-        NoStatusChange :
-            If the status does not actually change (e.g., the code is trying
-            to change it to a status it already has), this is raised. This
-            allows callers to catch and handle this case.
+        RuntimeError
+            Something went really wrong; you appear to have duplicate taskid
+            rows. This should never occur.
+        NoStatusChange
+            The update couldn't happen. This likely means that the database
+            allowed another process to mark the task as the updated status.
         """
-        ...
+        if result.rowcount > 1: # -no-cov-
+            raise RuntimeError("Database seems to have more than 1 row with"
+                               f"taskid '{taskid}'. This should not happen.")
+        elif result.rowcount == 0:
+            raise NoStatusChange(f"Task '{taskid}' could not change from "
+                                 f"{old_status} to {status}")
 
     def check_out_task(self):
```

Comment:
docstring

Reply:
I was hoping we could count on inheritance from the ABC instead of having multiple copies of the same string?